From 6d0394d27d570fe3586d5d25ec159e602496b1aa Mon Sep 17 00:00:00 2001
From: Gilles Sadowski <erans@apache.org>
Date: Fri, 9 Aug 2013 13:55:49 +0000
Subject: [PATCH] MATH-1020. Fixed "nextPermutation" in "RandomDataGenerator".
 Bug showed up when using a fixed version of "shuffle" (MATH-1019) added in
 "MathArrays" (MATH-1010). Added overloaded "shuffle" method (used in the
 above fix).

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/math/trunk@1512306 13f79535-47bb-0310-9956-ffa450edef68
---
 .../math3/random/RandomDataGenerator.java     | 10 ++++------
 .../apache/commons/math3/util/MathArrays.java | 20 +++++++++++++++++++
 2 files changed, 24 insertions(+), 6 deletions(-)

diff --git a/src/main/java/org/apache/commons/math3/random/RandomDataGenerator.java b/src/main/java/org/apache/commons/math3/random/RandomDataGenerator.java
index b662999604..fe6e72a1bf 100644
--- a/src/main/java/org/apache/commons/math3/random/RandomDataGenerator.java
+++ b/src/main/java/org/apache/commons/math3/random/RandomDataGenerator.java
@@ -46,6 +46,7 @@
 import org.apache.commons.math3.exception.NumberIsTooLargeException;
 import org.apache.commons.math3.exception.OutOfRangeException;
 import org.apache.commons.math3.exception.util.LocalizedFormats;
+import org.apache.commons.math3.util.MathArrays;
 
 /**
  * Implements the {@link RandomData} interface using a {@link RandomGenerator}
@@ -639,13 +640,10 @@ public int[] nextPermutation(int n, int k)
         }
 
         int[] index = getNatural(n);
-        shuffle(index, n - k);
-        int[] result = new int[k];
-        for (int i = 0; i < k; i++) {
-            result[i] = index[n - i - 1];
-        }
+        MathArrays.shuffle(index, getRandomGenerator());
 
-        return result;
+        // Return a new array containing the first "k" entries of "index".
+        return MathArrays.copyOf(index, k);
     }
 
     /**
diff --git a/src/main/java/org/apache/commons/math3/util/MathArrays.java b/src/main/java/org/apache/commons/math3/util/MathArrays.java
index 8123ae4cc7..01416ca8d9 100644
--- a/src/main/java/org/apache/commons/math3/util/MathArrays.java
+++ b/src/main/java/org/apache/commons/math3/util/MathArrays.java
@@ -1505,4 +1505,24 @@ public static void shuffle(int[] list,
             throw new MathInternalError(); // Should never happen.
         }
     }
+
+    /**
+     * Shuffle the entries of the given array.
+     *
+     * @param list Array whose entries will be shuffled (in-place).
+     * @param rng Random number generator.
+     */
+    public static void shuffle(int[] list,
+                               RandomGenerator rng) {
+        shuffle(list, 0, Position.TAIL, rng);
+    }
+
+    /**
+     * Shuffle the entries of the given array.
+     *
+     * @param list Array whose entries will be shuffled (in-place).
+     */
+    public static void shuffle(int[] list) {
+        shuffle(list, new Well19937c());
+    }
 }
